src/librustc/ty/util.rs

```diff
@@ -713,10 +716,17 @@ impl<'tcx> TyCtxt<'tcx> {
                 }
                 let substs = substs.fold_with(self);
                 if self.seen_opaque_tys.insert(def_id) {
-                    let generic_ty = self.tcx.type_of(def_id);
-                    let concrete_ty = generic_ty.subst(self.tcx, substs);
-                    let expanded_ty = self.fold_ty(concrete_ty);
-                    self.seen_opaque_tys.remove(&def_id);
+                    let expanded_ty = match self.expanded_cache.get(&def_id) {
```

Comment:
Why not use the `.entry` API?

Reply:
Yeah! Entry! Entry! I ❤️ Entry!

Reply:
Oh, I think I see why not -- the `self.fold_ty` call would error out with `entry`, wouldn't it?

Reply:
Yeah, the fact that the closure would need `&mut self` prevents me from using `entry` here :/